BUGFIX: The AssetUsageInNodePropertiesStrategy breaks for non-site-nodes

daniellienert opened the pull request:
The AssetUsageInNodePropertiesStrategy gathers the usage of assets
in site nodes. It uses the method
`NodePaths::getRelativePathBetween(SiteService::SITES_ROOT_PATH, $nodePath));`
to build the path between the site root and the node referencing the
asset. This method throws an exception when the asset is refernced
in a node wich is not under the site tree.

daniellienert commented:
@radmiraal Maybe the strategy should be renamed to AssetUsageIn**Site**NodePropertiesStrategy?

radmiraal commented:
hey @daniellienert, I'm not sure about what the original code was meant to do, just moved it from the asset controller. But only checking nodes below a site node sounds strange to me. For Neos context it might be right, but if we would do that we'd need a typocr strategy too.

My gut feeling says like this fixes the wrong issue or is at least incomplete

radmiraal commented:
> Maybe the strategy should be renamed to AssetUsageInSiteNodePropertiesStrategy?

Thinking about it then a better name would be AssetUsageInSiteStrategy because SiteNode would hint to the asset only being used in the root node of the site

daniellienert commented:
> But only checking nodes below a site node sounds strange to me. 

The `AssetUsageInNodePropertiesStrategy` definitely is in the neos context. It explicitly checks for TYPO3.Neos:Document and calculates a node path to the *sites* root node. My fix skips all findings that are not in a path under sites and so avoids the exception that is thrown when it tries to calculate the path to sites. So It might feel strange to only check the site nodes, but thats what this strategy is actually doing - wit the bug that it fails if a node not in the sites path is referencing an asset.

radmiraal commented:
Yes, I do agree, so probably renaming is one thing to do. But imho the strategy shouldn't fetch objects it wouldn't check. So the fix should then move to getRelatedNodes selecting only nodes below the SITE_ROOT_PATH. That could save quite a lot of nodes in memory.

The second issue is that by finding this bug you found a missing feature ;-) That's the fact we need a strategy in CR too maybe, but that could mean a dependency from CR to Media... so that's definitely something out of scope for the current PR. Maybe we should file it in jira though.

aertmann commented:
hmm shouldn't we make it work for nodes outside of a site instead? this results in unexpected behavior and don't see a good reason why it should work like that..

kitsunet commented:
Yes please.

radmiraal commented:
What is not really clear to me by this PR is what actually breaks. Thinking about it it has to be constructing the AssetUsageInNodeProperties object and then rendering the usage table.
I already made the documentNode argument of the AssetUsageInNodeProperties::__construct optional, we could do the same for the site node. Then it only needs an extra condition in the template to fix the rendering.

The only problem that would then happen (but that's a separate thing to do) is that the number of usages could differ from the number of rendered rows, but that's already a fact as we don't have a way to render custom strategies yet. Maybe we could just add a "and x other usages" below that table or so.

daniellienert commented:
Hey, 
Maybe it will help to to clear things, if I describe my scenario in detail.
I am currently working on meta data management for assets where I use the CR to store the data under a root node called "assets". Lately I added the asset object as a reference to the meta data which makes it very convenient to directly access the asset after querying the meta data object.

The meta data in CR is handled automatically - added, deleted and updated with the asset. In this scenario the CR is used as a technical data storage which is semantically completely different as having document and content nodes there. In this case, these nodes shouldn't be displayed to the editor, as he doesn't care about them. And definitely the paths don't make sense to an editor.

And as the CR is so flexible one could think about other scenarios, where it also doesn't make sens to display a node path for an asset usage,

So I vote for not make it more generic but, for make it more specific and only display usages in nodes that derive from Document or Content, as these are currently the only things, the UI can provide something to the user to take action. As the thing is called "Strategy" - it enables us to build more strategies, for other asset references - als of other references from the CR.

I hope I made things more clear (alway hard in english) - I hope I make it to document the meta data packages and put them out of the shade soon, so we all can have the same picture.

radmiraal commented:
@daniellienert thanks for explaining, and that is exactly what the strategies should allow and what I meant with the fact we miss a strategy in CR now. While working on the feature I also thought about how to render usages 'outside' of Neos, but at that point that was not within the scope of the task.

So I see the usecase, but the problem that arises is that it will be fairly complicated to implement. Like adding the strategy to CR (that might be the most ideal and clean place for your usecase) would add a dependency to the Media package in CR which I would not be happy with at this time.

What we could do is add a check to the relatedNodesAction and construct a list of 'renderable' usages, being the ones of type AssetUsageInNodeProperties with a site node / document node attached. This way we could have the rendered list as is now, and implement this with the least effort.

Btw: I do not agree to the "editor does not care about them", the asset is locked as being in used and it will be shown in the usage count. Seeing just 1 rendered item and seeing "the asset is used 20 times" is kind of bad imho.

If we change the Neos strategy to handle this usecase we prevent creating a strategy in CR. And Neos is in the end the package where the media package and CR come to play together.

If you like I can try to extend the PR to do what I try to explain, I think it suits your usecase :-)

daniellienert commented:
@radmiraal 
> the asset is locked as being in used and it will be shown in the usage count. 

With my change, It isn't shown in the usage count, and the "locking" didn't seem to work without this PR for non site nodes either.
I added you to my repo in case you want to work on this PR directly. 

Here are the packages listed to test the scenario: https://discuss.neos.io/t/rfc-adding-additional-properties-to-media-assets/572/25

> If you like I can try to extend the PR to do what I try to explain, I think it suits your usecase :-)

My main acceptance criteria is simple - no exception when clicked on image detail with asset references in non site-root nodes. :)

hlubek commented:
I stumbled into this issue too. My preference would be to throw an exception for "invalid nodes" where a missing site could be one of the cases. The `\TYPO3\Neos\Domain\Strategy\AssetUsageInNodePropertiesStrategy::getUsageReferences` method could then skip nodes that are invalid.

radmiraal commented:
@hlubek I'm working on a fix, hope to push it this week. It won't skip invalid nodes though as that would fix the symptom and break the actual goal of the strategies.

hlubek commented:
@radmiraal Okay, good to know and even better ;)

daniellienert commented:
:mega: Hey @radmiraal any news on this? :)

kdambekalns commented:
So @radmiraal informed me that he has this still on the agenda. Try harder to coordinate things, fellow Neos people. Ok? ;)

daniellienert commented:
👍

radmiraal commented:
I'll try to fix this asap now, completely lost it in my head :(

radmiraal commented:
@daniellienert hey, I pushed a commit to https://github.com/radmiraal/neos-development-collection/commit/69207290cbe7cc02701fc933e41bfc59229daf41 that shows what I meant. While working on it I see that there're some naming things to think about, and also I'd like to see how we can improve the rendering of that nodeUsage table. 

From the beginning that table was only created with site nodes in mind, but I think the general idea of the change should be clear ;-) It would be nice if people can have a look at that change and give feedback if that's roughly a feasible change and give some suggestions how to fix the naming issue with the UsageReferences and the structure of the array that is used for rendering the relatedNodes.

kitsunet commented:
@daniellienert @radmiraal can you two coordinate this and make it merge ready? Thx!

daniellienert commented:
@radmiraal First short response - I am really busy ATM, but I am on it - I tested your commit and found no issues so far. Will have a deeper look at the code next.
